pkg/webhooks/sdnmigration/sdnmigration.go

```diff
@@ -0,0 +1,170 @@
+	admissionregv1 "k8s.io/api/admissionregistration/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	admissionctl "sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"github.com/openshift/managed-cluster-validating-webhooks/pkg/webhooks/utils"
+)
+
+const (
+	WebhookName               string = "sdn-migration-validation"
+	privilegedServiceAccounts string = `^system:serviceaccounts:(kube.*|openshift.*|default|redhat.*|osde2e-[a-z0-9]{5})`
+	privilegedUsers           string = `system:admin`
+	docString                 string = `Managed OpenShift customers may not modify the network config type because it can can degrade cluster operators and can interfere with OpenShift SRE monitoring.`
+)
+
+var (
+	log                         = logf.Log.WithName(WebhookName)
+	privilegedServiceAccountsRe = regexp.MustCompile(privilegedServiceAccounts)
+	privilegedUsersRe           = regexp.MustCompile(privilegedUsers)
+
+	scope = admissionregv1.ClusterScope
+	rules = []admissionregv1.RuleWithOperations{
+		{
+			Operations: []admissionregv1.OperationType{"UPDATE"},
+			Rule: admissionregv1.Rule{
+				APIGroups:   []string{"config.openshift.io"},
+				APIVersions: []string{"*"},
+				Resources:   []string{"networks"},
+				Scope:       &scope,
+			},
+		},
+	}
+)
+
+type NetworkConfigWebhook struct {
+	mu sync.Mutex
+	s  runtime.Scheme
+}
+
+// Authorized will determine if the request is allowed
+func (w *NetworkConfigWebhook) Authorized(request admissionctl.Request) (ret admissionctl.Response) {
+	ret = admissionctl.Denied("Changing the network type is not allowed")
+	ret.UID = request.AdmissionRequest.UID
+
+	// allow if modified by an allowlist-ed service account
+	for _, group := range request.UserInfo.Groups {
+		if privilegedServiceAccountsRe.Match([]byte(group)) {
+			ret = admissionctl.Allowed("Privileged service accounts may access")
+			ret.UID = request.AdmissionRequest.UID
+		}
+	}
+
+	if request.Operation == admissionv1.Update {
+		decoder, err := admissionctl.NewDecoder(&w.s)
+		if err != nil {
+			log.Error(err, "failed to render a network config from request.Object")
+			ret = admissionctl.Errored(http.StatusBadRequest, err)
+			ret.UID = request.AdmissionRequest.UID
+		}
+
+		object := &configv1.Network{}
+		oldObject := &configv1.Network{}
+
+		if err := decoder.DecodeRaw(request.Object, object); err != nil {
+			log.Error(err, "failed to render a Network from request.Object")
+			ret = admissionctl.Errored(http.StatusBadRequest, err)
+			ret.UID = request.AdmissionRequest.UID
+		}
+		if err := decoder.DecodeRaw(request.OldObject, oldObject); err != nil {
+			log.Error(err, "failed to render a Network from request.OldObject")
+			ret = admissionctl.Errored(http.StatusBadRequest, err)
+			ret.UID = request.AdmissionRequest.UID
+		}
+
+		if object.Spec.NetworkType != oldObject.Status.NetworkType {
```

Comment:
```suggestion
		if object.Spec.NetworkType != oldObject.Spec.NetworkType {
```
Is there a reason to not compare `.spec.networkType` here? (Mostly out of curiosity)

Reply:
I was taking a cue from [the configv1 docs](https://pkg.go.dev/github.com/openshift/api/config/v1#Network) where it suggests reading from `Status` instead of `Spec`

> 	// spec holds user settable values for configuration.
	// As a general rule, this SHOULD NOT be read directly. Instead, you should
	// consume the NetworkStatus, as it indicates the currently deployed configuration.
	// Currently, most spec fields are immutable after installation. Please view the individual ones for further details on each.

Reply:
Awesome, thanks for the explanation!